src/Executables/CombineH5/CombineH5.cpp

```diff
@@ -0,0 +1,144 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include <boost/program_options.hpp>
+#include <cstddef>
+#include <map>
+#include <string>
+#include <vector>
+
+#include "DataStructures/DataVector.hpp"
+#include "IO/H5/AccessType.hpp"
+#include "IO/H5/File.hpp"
+#include "IO/H5/SourceArchive.hpp"
+#include "IO/H5/VolumeData.hpp"
+#include "Parallel/Printf.hpp"
+#include "Utilities/FileSystem.hpp"
+
+// Charm looks for this function but since we build without a main function or
+// main module we just have it be empty
+extern "C" void CkRegisterMainModule(void) {}
+
+// Compares source archives of each file against the 0th volume file. Returns
+// true if each source archive matches the 0th one
+bool check_src_files(const std::vector<std::string>& file_names) {
+  h5::H5File<h5::AccessType::ReadWrite> initial_file(file_names[0], true);
+  const auto& src_archive_object_initial =
+      initial_file.get<h5::SourceArchive>("/src");
+  const std::vector<char>& src_tar_initial =
+      src_archive_object_initial.get_archive();
+  for (size_t i = 1; i < file_names.size(); ++i) {
+    h5::H5File<h5::AccessType::ReadWrite> comparison_file(file_names[i], true);
+    const auto& src_archive_object_compare =
+        comparison_file.get<h5::SourceArchive>("/src");
+    const std::vector<char>& src_tar_compare =
+        src_archive_object_compare.get_archive();
+
+    if (src_tar_initial != src_tar_compare) {
+      return false;
+    }
+  }
+  return true;
+}
+
+void combine_h5(const std::string& file_prefix, const std::string& subfile_name,
+                const std::string& output) {
+  // Parses for and stores all input files to be looped over
+  const std::vector<std::string>& file_names =
+      file_system::glob(file_prefix + "*.h5");
+
+  // Checks that volume data was generated with identical versions of SpECTRE
+  if (!check_src_files(file_names)) {
+    ERROR(
+        "One or more of your files were found to have differing src.tar.gz "
+        "files, meaning that they may be from differing versions of SpECTRE.");
+  }
+
+  // Instantiates the output file and the .vol subfile to be filled with the
+  // combined data later
+  h5::H5File<h5::AccessType::ReadWrite> new_file(output + "0.h5", true);
+  new_file.insert<h5::VolumeData>("/" + subfile_name + ".vol");
+  new_file.close_current_object();
+
+  // std::map to store the sorted volume data
+  std::map<int, std::vector<
+                    std::tuple<size_t, double, std::vector<ElementVolumeData>>>>
+      file_map;
+
+  // Opens the original files, sorts the volume data, and stores it in the
+  // std::map
+  for (size_t i = 0; i < file_names.size(); ++i) {
```

Comment:
This ends up storing an enormous amount of data in memory. Is it possible to adjust the algorithm to only need to store one time (observation ID) in memory at a time? E.g. for binary merger files we could quite reasonably end up with hundreds of gigabytes of data, and we cannot store that in memory. We hopefully won't need to handle one component at a time, but I think that can safely be a "future problem"

E.g. something like using `list_observation_ids()`, then loop over each one, for each one `get_observation_value` then get the data `get_data_by_element(time * (1.0 - 4.0 * std::numeric_limits<double>::epsilon()), time * (1.0 + 4.0 * std::numeric_limits<double>::epsilon()))`  (I chose to, fairly arbitrarily, bound the time by 4*epsilon), combine it, and write it.